nemo/collections/asr/modules/conformer_encoder.py

```diff
@@ -235,57 +321,116 @@ def set_max_audio_length(self, max_audio_length):
             self.register_buffer('seq_range', seq_range, persistent=False)
         self.pos_enc.extend_pe(max_audio_length, device)
 
+        att_mask = torch.ones(1, max_audio_length, max_audio_length, dtype=torch.bool, device=device)
+        if self.chunk_size is None:
+            if self.att_context_size[0] >= 0:
+                att_mask = att_mask.triu(diagonal=-self.att_context_size[0])
+            if self.att_context_size[1] >= 0:
+                att_mask = att_mask.tril(diagonal=self.att_context_size[1])
+        else:
+            chunk_idx = torch.arange(0, max_audio_length, dtype=torch.int, device=att_mask.device)
+            chunk_idx = torch.div(chunk_idx, self.chunk_size, rounding_mode="trunc")
+            diff_chunks = chunk_idx.unsqueeze(1) - chunk_idx.unsqueeze(0)
+            chunked_limited_mask = torch.logical_and(
+                torch.le(diff_chunks, self.left_chunks_num), torch.ge(diff_chunks, 0)
+            )
+            att_mask = torch.logical_and(att_mask, chunked_limited_mask.unsqueeze(0))
+
+        if hasattr(self, 'att_mask'):
+            self.att_mask = att_mask
+        else:
+            self.register_buffer('att_mask', att_mask, persistent=False)
+
     @typecheck()
-    def forward(self, audio_signal, length=None):
+    def forward(self, audio_signal, length=None, cache_last_channel=None, cache_last_time=None):
         self.update_max_seq_length(seq_length=audio_signal.size(2), device=audio_signal.device)
-        return self.forward_for_export(audio_signal=audio_signal, length=length)
+        return self.forward_for_export(
+            audio_signal=audio_signal,
+            length=length,
+            cache_last_channel=cache_last_channel,
+            cache_last_time=cache_last_time,
+        )
 
     @typecheck()
-    def forward_for_export(self, audio_signal, length):
+    def forward_for_export(self, audio_signal, length, cache_last_channel=None, cache_last_time=None):
         max_audio_length: int = audio_signal.size(-1)
-
         if max_audio_length > self.max_audio_length:
             self.set_max_audio_length(max_audio_length)
-
         if length is None:
             length = audio_signal.new_full(
                 audio_signal.size(0), max_audio_length, dtype=torch.int32, device=self.seq_range.device
             )
 
+        if cache_last_channel is not None:
+            cache_last_time_next = torch.zeros_like(cache_last_time)
+        else:
+            cache_last_time_next = None
         audio_signal = torch.transpose(audio_signal, 1, 2)
 
         if isinstance(self.pre_encode, nn.Linear):
-            audio_signal = self.pre_encode(audio_signal)
+            audio_signal = self.pre_encode(x=audio_signal)
         else:
-            audio_signal, length = self.pre_encode(audio_signal, length)
+            audio_signal, length = self.pre_encode(x=audio_signal, lengths=length)
+            if self.streaming_cfg.drop_extra_pre_encoded > 0 and cache_last_channel is not None:
```

Comment:
Where is self.streaming_cfg set?

Reply:
Nevermind, saw below. Add a comment stating this config is set via the setup_streaming_cfg() method. - or move the streaming config method below the constructor.

Reply:
added the comment.